### PR TITLE
Fixes #24 Update WebSocket protocol to “BLIP_3"

### LINKED ISSUE #24: Update WebSocket protocol to “BLIP_3"
The `Sec-WebSocket-Protocol` value being sent in the WebSocket handshake is currently `BLIP_3a2`. For the release we should change it to `BLIP_3`, both because it's the final version 3 of BLIP, and to prevent anyone trying to use the GA release of CBL with a beta of SG or vice versa.

(The flip side of this is couchbase/couchbase-lite-core#411. Both commits will need to be pushed at the same time, of course.)

### PR DESCRIPTION
**Pull request by tleyden**
## Testing

- [ ] Added unit test `TestIncludesProtocol` to test `includesProtocol()` in isolation
- [ ] Test the litecore blip replication test "API Continuous Push" after modifying `kWSProtocolName` to be `BLIP_3` and `BLIP_3+CBMobile_2` (in `BLIPConnection.hh`)

### CODE REVIEW

File: context.go [hunk 1]
```diff
@@ -201,13 +208,38 @@ func (context *Context) logFrame(format string, params ...interface{}) {
...
+// "WebSocketSubProtocol1, WebSocketSubProtocol2+CBMobile_2"
+//
+// includes at least one protocol with a prefix equivalent to the protocolPrefix passed in.
+// For example, if the protocolPrefix is BLIP_3, then this would be a valid header:
+//
+// "BLIP_3+whatever"
+//
+// Returns a boolean indicating if any of the subprotocols listed in the header string matched.  If so,
+// it returns the matching subprotocol.
+func includesProtocolWithPrefix(header string, protocolPrefix string) (matched bool, matchedSubprotocol string) {
```
Comment (snej) [inline context.go]: This isn't quite what I intended for the "+CBMobile_2" part. It refers to the replication protocol (and version), to distinguish it from some other usage of BLIP. That way if a client connects asking for protocol "BLIP_3+MultiPlayerGame" you can reject it because it's not a replication client. Or if a future version of Couchbase Lite sends "BLIP_3+CBMobile_3", you can reject it because you don't know that replication protocol.

In other words, the Context should still compare the protocol string exactly. But the required protocol string needs to be configurable since go-blip shouldn't hardcode the "+CBMobile_2" part.

Comment (tleyden) [reply context.go]: @snej ok makes sense.  I pushed up a new PR, take a look and let me know if that's more along the lines of what you were thinking.

File: protocol.go [hunk 2]
```diff
@@ -1,9 +1,5 @@
 package blip
 
-// WebSocket [sub]protocol name for BLIP, used during WebSocket handshake.
-// Client request must indicate that it supports this protocol, else the handshake will fail.
-const WebSocketProtocolName = "BLIP_3a2"
```
Comment (snej) [inline protocol.go]: I think it's best to keep this, since every sub-protocol used by a caller should begin with this string, rather than hardcoding the BLIP protocol version into the caller.

**Comment by tleyden** [comment]
@snej reworked accordingly.  Now the `NewContext()` only takes the "application protocol id", and constructs the full websocket subprotocol from that by adding it to the (fixed) `BLIP_3` protocol identifier.  See latest round of changes.

File: context.go [hunk 3]
```diff
@@ -24,6 +24,11 @@ func Unhandled(request *Message) {
 
 // Defines how incoming requests are dispatched to handler functions.
 type Context struct {
+
+	// The WebSocket subprotocol that this blip context is constrained to.  Eg: BLIP_3+CBMobile_2
+	// Client request must indicate that it supports this protocol, else WebSocket handshake will fail.
+	WebSocketSubProtocol string
+
 	HandlerForProfile map[string]Handler // Handler function for a request Profile
```
Comment (snej) [inline context.go]: OMG the types don't line up! ;) Need to run `go format`.

File: context.go [hunk 4]
```diff
@@ -47,15 +52,20 @@ type LogContext interface {
 //////// SETUP:
 
 // Creates a new Context with an empty dispatch table.
-func NewContext() *Context {
-	return NewContextCustomID(fmt.Sprintf("%x", rand.Int31()))
+func NewContext(AppProtocolId string) *Context {
```
Comment (snej) [inline context.go]: Shouldn't this be lowercase? It's a function parameter. (Ditto for line 63.)

File: context.go (hunk 3 above)
Comment (tleyden) [reply context.go]: Haha, actually this is what `gofmt` does when there is an empty space between "sections" in a struct.  I wanted to more comments on `WebSocketSubProtocol` than would fit to the right of the field name, so I put it into it's own section.

File: context.go (hunk 4 above)
Comment (tleyden) [reply context.go]: Thanks for catching.  Fixed and pushed.
